[![DOI](https://img.shields.io/badge/DOI-10.1038%2Fs41525--022--00292--2-green)](https://doi.org/10.1038/s41525-022-00292-2)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5846705.svg)](https://doi.org/10.5281/zenodo.5846705)
[![Support this project by running your production jobs at BatchX](https://images.batchx.io/gh-badge-logo.svg)](https://platform.batchx.io/uniovi/profile "Support this project by running your production jobs at BatchX")

# Context 

Mappability can be challenging in repetitive regions when analyzing NGS-derived data. Local **duplication of genomic** regions hampers the alignment of reads in those parts of the genome, since it **prevents unambiguously** **alingning reads**. Therefore, those reads are usually assigned a random location (among the possible ones) and a low _mapping quality_. Since reads are assigned a random location, **mutations** **are usually either** **missed** (mutated reads are dispersed among all possible alignment regions) or, even worse, **called multiple times** leading to the rise of false positives (enough reads to call a variant in multiple alignment regions). For this reason, **low mapping quality regions are often ignored** in variant calling pipelines.

__Armadillo__ is a somatic variant calling pipeline that aims at repetitive regions for tumor-control paired samples, in order to recover mutations that are systematically missed. In order to avoid both false positives and false negatives, it first __finds all candidate regions__ for a repetitive region of interest (RROI). Then, it extracts all reads coming from all regions and aligns them at the RROI sequence only instead of the whole genome, forcing all reads to be aligned together. This way, dilution of variant-supporting reads is prevented, as well as the false positives caused by mistakenly call a mutation at multiple copies. 
Finally, the variant calling step is carried out. After usual __heuristic filters__, we use __context-specific filters__ to prevent false positives related the stacking step and, lastly, run a __CNN model to classify the mutations__.

## In-depth review of Armadillo

As stated before, calling mutations at repetitive region can be risky, since we'll probably find false positives. In order to prevent this, Armadillo will:

1. **Retrieve all copies of a repetitive region of interest** (RROI). We just need to pass a list of regions of interest and it will find the other copies of each region we are interested in.
2. **Extract the reads** **aligned at each locus** and align them against the sequence of our RROI (instead of the whole genome). This way, **the alignment will be unambiguous** since all reads that could be aligned at each copy will be forced to be aligned at a single locus.
3. Apply **heuristic filters** to select candidate variants (at least 3 variant-supportive reads in tumor, less than 3 variant-supportive reads in control sample to remove germinal variants, minimum coverage...)
4. Apply **specific filters**. The step 2 assures that all possible reads are being used, increasing sensitivity, but it also leads to an increase of false positives (artifacts arise). It checks that nearby variants (**context**) **match** in all mutated reads, it tries to **find non-mutated reads that carry the same context** in the control sample..., so false positives are controlled. 
5. Perform **MonteCarlo simulations** to compute the chance of our mutation being somatic or germinal. 
6. Run a **CNN model** trained with curated mutations from our paper that takes into account the context for calling the mutation.
7. Make sure that the same mutation isn't returned multiple times. If a gene has two copies (gene1, gene2) and both gene1 and gene2 have been used as input, only the one will be returned in the final file.

More information on the pipeline can be retrieved in our [GitHub repository](https://github.com/xa-lab/armadillo) or in [our paper](https://www.nature.com/articles/s41525-022-00292-2).

# Inputs

## Required inputs

1. `name`: Experiment name
2. `controlGenome`: Control sample genome
3. `tumorGenome`: Tumor sample genome

## Optional inputs

1. `controlIndex` and `tumorIndex`: BAM index generated with `samtools`. Recommended for shorter runs
2. `armadilloData`: armadillo data-prep output compressed as tar.gz
3. `referenceVersion`: Select reference genome. If custom, armadilloData output of armadillo data-prep script must be included as tar.gz with the armadilloData argument. Default is hg38.
4. `roisList`: List of ROIs to analyze instead of the whole list in armadillo_data (BED format or chr:start-end format accepted). Header not needed.
7. `returnBams`: Return minibams generated by armadillo
8. `returnDataprep`: Return data-prep results (useful if same analysis will be done on a batch of samples)
9. `returnDiscarded`: Return discarded variants

> __Note__: Either `armadilloData` or `roisList` is needed. Also, `genomeRef` is required if `armadilloData` is not provided.

## Additional parameters


|     Parameter     |                             Description                            | Default |
|:-----------------:|:------------------------------------------------------------------:|:-------:|
|  controlCoverage |                   Coverage of the control sample                   |    30   |
|   tumorCoverage  |                    Coverage of the tumor sample                    |    30   |
| controlThreshold | Maximum alternative-allele supportive reads allowed in the control |    3    |
|  tumorThreshold  |      Minimum coverage required for a variant to be considered      |    6    |
|    baseQuality   |         Minimum base quality required for the tumor genome         |    30   |
|        mapq       |            Minimum MapQ for reads after being collapsed            |    30   |
|      GCcutoff     |                         Maximum GC% allowed                        |    80   |

# Outputs

1. `vcf`: VCF with the mutations detected by the pipeline
2. `tumorMinibam` and `controlMinibam`: minibams generated by the pipeline (_optional_, only if `returnBams=true`)
3. `tumorMinibamIdx` and `controlMinibamIdx`: minibam indexes (_optional_, only if `returnBams=true`)
4. `discardedVariants`: log file with all discarded candidates (_optional_, only if `returnDiscarded=true`)

# Example 

1. Copy bams and ROIs  to batchX
```bash
bx cp tumor.bam control.bam rois bx://test/
```

> A list of precomputed rois is available in our github repo for [hg19](https://github.com/pbousquets/armadillo-batchx/tree/master/lib/armadillo_data_hg19/rois) or [here](https://github.com/pbousquets/armadillo-batchx/tree/master/lib/armadillo_data_hg38/rois) for hg38. If a concrete roi isn't in those lists, try to run armadillo-data prep from our [original repository](https://github.com/xa-lab/armadillo), or ask for it in a [pull request](https://github.com/pbousquets/armadillo-batchx/pulls).

3. Submit job

```bash
bx submit -v=10 -m=16000 uniovi@labxa/armadillo:1.0.0 \
'{
  "name": "test",
  "controlGenome": "/test/control.bam",
  "tumorGenome": "/test/tumor.bam",
  "referenceVersion": "hg38"}'
```

# FAQ

- **What is a BAM? What is the index?**

A BAM file is the output we get after aligning the reads against a reference genome. This step is absolutely required, since this type of files report not only the sequences but also information about the location of the reads in the reference genome and mappability (mapQ). The index is not required, but __recommended__. The indexes allow the pipeline to retrieve information of a region directly, instead of reading the whole file (very time-consuming). If not provided, the pipeline will index them automatically, but it will take some minutes.

If your sequences are in FASTQ or FASTA format, we recommend to align them with [BWA](https://platform.batchx.io/batchx/images/bioinformatics%2Fbwa%2Fmem/2.0.2). Also, the BAM indexes can be created with [SAMtools index](https://platform.batchx.io/batchx/images/bioinformatics%2Fsamtools%2Findex/1.0.0#documentation).

- **How do I get a file of regions of interest?**

Armadillo doesn't analyze the whole genome as regular variant callers. It relies on a preset of regions that may or may not be repetitive, but it actually needs to be told where to find the mutations. Then, it will check which of those regions are repetitive, and will call mutations exclusively in those. A simple tabulated file with chromosome, start and end is enough. It also accepts chr:start-end format.  


- **Does the final VCF need any postprocessing?**

No, it does not. Nonetheless, if you are using a batch of samples, using a panel of normals can be helpful to reduce the noise. Panel of normals are collections of variants that appear in multiple healthy samples. If a somatic mutation exists in those panels, it's not expected to be a cancer driver. Also, it can help removing some unexpected false positives. In order to do so with our data, use `returnDiscarded: true` in your input files. This way, a list of all discarded mutations will be returned. Then, you can merge all `discarded_variant.log` files and keep all variants with the flag `germline_change`. This way, a list of variants found in the healthy sample will be returned. After that, count the occurrences of each mutation. Short example in linux:
```
cd $my_results # Directory where all armadillo results are saved
cat */discarded_variants.log | grep germline_change | cut --complement  -f3 | sort | uniq -c | awk '{first = $1; $1=""; print $0, first}' > mypanel.txt
```
Now, this database can be used to know if a variant appears recurrently in healthy samples.

- **I have Whole Exome Sequencing derived data, can I use Armadillo as well?**

Although there are no restrictions in Armadillo, there are two issues that require attention: 
1. WES libraries are focused in protein-coding genes and tend to avoid problematic regions, like those with high GC content or those that are repetitive. 
2. Armadillo has been developed on a WGS basis. It hasn't been tested on WES data because of what we explained in the previous point.

Thus, __we do not recommend to use Armadillo for WES data__.

# Tools version

* [BWA](http://bio-bwa.sourceforge.net/) (built with v.0.7.17)
* [Samtools](https://www.htslib.org/doc/1.9/samtools.html) (built with v.1.9)
* [Python3](https://www.python.org) (built with v.3.7)

# Authors

* **Pablo Bousquets Muñoz** - bousquetspablo@uniovi.es
* **Ander Díaz Navarro**
* **Xose Antón Suárez Puente**
